srclib/src/locator.rs

```diff
@@ -0,0 +1,491 @@
+    /// - `{fetcher}+{project}`
+    /// - `{fetcher}+{project}$`
+    /// - `{fetcher}+{project}${revision}`
+    ///
+    /// Projects may also be namespaced to a specific organization;
+    /// in such cases the organization ID is at the start of the `{project}` field
+    /// separated by a slash. The ID can be any non-negative integer.
+    /// This yields the following formats:
+    /// - `{fetcher}+{org_id}/{project}`
+    /// - `{fetcher}+{org_id}/{project}$`
+    /// - `{fetcher}+{org_id}/{project}${revision}`
+    ///
+    /// This parse function is based on the function used in FOSSA Core for maximal compatibility.
+    ///
+    /// This implementation ignores the `revision` segment if it exists. If this is not preferred, use [`Locator`] instead.
+    pub fn parse(locator: &str) -> Result<Self, Error> {
+        let full = Locator::parse(locator)?;
+        Ok(Self {
+            fetcher: full.fetcher,
+            org_id: full.org_id,
+            project: full.project,
+        })
+    }
+
+    /// Promote a `PackageLocator` to a [`Locator`] by providing the value to use for the `revision` component.
+    pub fn promote(self, revision: Option<String>) -> Locator {
+        Locator {
+            fetcher: self.fetcher,
+            org_id: self.org_id,
+            project: self.project,
+            revision,
+        }
+    }
+}
+
+impl Display for PackageLocator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let converted = Locator::from(self);
+        write!(f, "{converted}")
+    }
+}
+
+impl<'de> Deserialize<'de> for PackageLocator {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let raw = String::deserialize(deserializer)?;
+        PackageLocator::parse(&raw).map_err(serde::de::Error::custom)
+    }
+}
+
+impl Serialize for PackageLocator {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+/// [`Locator`] is closely tied with the concept of Core's "fetchers",
+/// which are asynchronous jobs tasked with downloading the code
+/// referred to by a [`Locator`] so that Core or some other service
+/// may analyze it.
+///
+/// `Fetcher` enumerates the identifiers of possible fetchers in Core.
+/// The intention is not to maintain the list of _all possible fetchers_;
+/// instead the intention is to provide a list of fetchers
+/// with which services in this repository interact.
+///
+/// For this reason this enum is marked non-exhaustive and always will be:
+/// it'll always be possible for fetchers to exist in Core or elsewhere
+/// before they're enumerated here (by choice or by accident).
+///
+/// For more information on the background of `Locator` and fetchers generally,
+/// refer to [Fetchers and Locators](https://go/fetchers-doc).
```

Comment:
This links to a private doc. I think that's probably okay — the link is still useful for people who have access to the doc — but wanted to call it out

Reply:
Yeah, that's one reason I left it a golink as well. I want it there for us to reference, but I think we can't make that doc public.